### Log-Loss (логарифмическая функция потерь)

$$\large logloss = - \frac{1}{l} \cdot \sum_{i=1}^l (y_i \cdot log(\hat y_i) + (1 - y_i) \cdot log(1 - \hat y_i))$$

здесь $\hat y$ — это ответ алгоритма на $i$-ом объекте, $y$ — истинная метка класса на $i$-ом объекте, а $l$ размер выборки.

Интуитивно можно представить минимизацию log-loss как задачу максимизации accuracy путем штрафа за неверные предсказания. Log-loss крайне сильно штрафует за уверенность классификатора в неверном ответе.

In [7]:
import numpy as np

def logloss(y_true, y_pred):

    return - (y_true * np.log(y_pred) + (1 - y_true) * np.log(1 - y_pred))


print('Logloss при неуверенной классификации %f' % logloss(1, 0.5))

print('Logloss при уверенной классификации и верном ответе %f' % logloss(1, 0.9))

print('Logloss при уверенной классификации и НЕверном ответе %f' % logloss(1, 0.1))

Logloss при неуверенной классификации 0.693147
Logloss при уверенной классификации и верном ответе 0.105361
Logloss при уверенной классификации и НЕверном ответе 2.302585


### Precision, recall & confusion matrix для многоклассовой классификации

<img src="CM1.png">

<img src="CM2.png">

Для класса Cats:
    
    precision = доля правильно предсказанных Cats / все предсказанные Cats = 4/13
    
    recall = доля правильно предсказанных Cats / все истинные Cats = 4/6

In [9]:
from sklearn import metrics

# Constants
C="Cat"
F="Fish"
H="Hen"

# True values
y_true = [C,C,C,C,C,C, F,F,F,F,F,F,F,F,F,F, H,H,H,H,H,H,H,H,H]
# Predicted values
y_pred = [C,C,C,C,H,F, C,C,C,C,C,C,H,H,F,F, C,C,C,H,H,H,H,H,H]

# Print the confusion matrix
print(metrics.confusion_matrix(y_true, y_pred))

# Print the precision and recall, among other metrics
print(metrics.classification_report(y_true, y_pred, digits=3))

[[4 1 1]
 [6 2 2]
 [3 0 6]]
              precision    recall  f1-score   support

         Cat      0.308     0.667     0.421         6
        Fish      0.667     0.200     0.308        10
         Hen      0.667     0.667     0.667         9

   micro avg      0.480     0.480     0.480        25
   macro avg      0.547     0.511     0.465        25
weighted avg      0.581     0.480     0.464        25



### F1-score

***F1-score = 2 × (precision × recall)/(precision + recall)***

F1-score - это среднее гармоническое точности и полноты

### Macro avg (average)

***macro average (precision) = среднее арифметическое всех (precision)***

Пример:
***macro avg precision = (0.308+0.667+0.667)/3=0.547***

### Micro avg (average)

***micro average (precision) = precision, но вычисленная на всех данных вместе.***

Пример:
precision = TP/(TP+FP)

TP - все правильно предсказанные объекты: 
$$TP = 4 + 2 + 6 = 12$$

FP - например, все Cat, предсказанные как Fish и т.д.
$$FP = 6 + 3 + 1 + 0 + 1 + 2 = 13$$

Поэтому
***micro avg precision*** $ = 12/(12+13)=0.480$

### Cтатьи про метрики качества многоклассовой классификации:
    
1. https://towardsdatascience.com/multi-class-metrics-made-simple-part-i-precision-and-recall-9250280bddc2
        
2. https://towardsdatascience.com/multi-class-metrics-made-simple-part-ii-the-f1-score-ebe8b2c2ca1